components/cam/src/physics/cam/shoc.F90

```diff
@@ -0,0 +1,2181 @@
+  real(r8), intent(out) :: brunt(shcol,nlev) 
+  ! SHOC mixing length [m]
+  real(r8), intent(out) :: shoc_mix(shcol,nlev) 
+
+! LOCAL VARIABLES
+  integer i, j, k, kk, kt
+  integer kl, ku, kb, kc, dothis, kli, kui
+  real(r8) :: deep_thresh, deep_thick, cloud_thick, lstarn, thresh
+  real(r8) :: cldmix, vonk, thedel, depth
+  real(r8) :: omn, betdz, bbb, term, qsatt, dqsat, bet
+  real(r8) :: thv_up, thv_dn, thedz, tscale, thefac, thecoef, thegam, norm
+  real(r8) :: stabterm, conv_var, tkes, mmax, cldthresh
+  logical lf, indexr
+  logical doclouddef
+  real(r8) :: conv_vel(shcol,nlev)
+  real(r8) :: thv_zi(shcol,nlevi)
+  
+  real(r8) :: numer(shcol)
+  real(r8) :: denom(shcol)
+  real(r8) :: cldarr(shcol) 
+  real(r8) :: l_inf(shcol)
+  real(r8) :: brunt2(shcol,nlev)
+ 
+  doclouddef = .true.
+ 
+  vonk=0.35_r8   ! Vonkarman constnt
+  tscale=400._r8 ! time scale set based on similarity results
+  brunt2(:,:) = 0._r8
+
+  ! Interpolate virtual potential temperature onto interface grid
+  call linear_interp(zt_grid,zi_grid,thv,thv_zi,nlev,nlevi,shcol,0._r8)
+
+  ! Define the brunt vaisalia frequency 
+  do k=1,nlev
+    do i=1,shcol
+      brunt(i,k) = (ggr/thv(i,k)) * (thv_zi(i,k+1) - thv_zi(i,k))/dz_zt(i,k) 
+    enddo
+  enddo 
+ 
+  ! Find length scale outside of clouds
+  do k=1,nlev
+    do i=1,shcol
+    
+      if (cldin(i,k) .eq. 0 .or. .not. doclouddef) then
+        tkes=sqrt(tke(i,k))
+	numer(i)=numer(i)+tkes*zt_grid(i,k)*dz_zt(i,k)
+	denom(i)=denom(i)+tkes*dz_zt(i,k)
+      else
+        cldarr(i)=1
+      endif
+    
+    enddo
+  enddo
+  
+  do i=1,shcol
+    if (denom(i) .gt. 0._r8) then
+      l_inf(i)=0.1_r8*(numer(i)/denom(i))
+    else
+      l_inf(i)=100._r8
+    endif
+  enddo
+ 
+  do k=1,nlev
+    do i=1,shcol
+    
+      tkes=sqrt(tke(i,k))
+      
+      if (brunt(i,k) .ge. 0) brunt2(i,k) = brunt(i,k)
+      
+      ! End of computation of local stability
+      if (k .eq. 1) then ! point closest to surface
+        term=600._r8 * tkes
+	shoc_mix(i,k)=term+(0.4_r8*zt_grid(i,k)-term)*exp(-zt_grid(i,k)/100._r8)
+      else
+        shoc_mix(i,k)=min(maxlen,(2.8284_r8*sqrt(1._r8/((1._r8/(tscale*tkes*vonk*zt_grid(i,k))) &
+	  +(1._r8/(tscale*tkes*l_inf(i)))+0.01_r8*(brunt2(i,k)/tke(i,k)))))/0.3_r8)  
+      endif      
+      
+    enddo  ! end i loop (column loop)
+  enddo ! end k loop (vertical loop)
+  
+  ! Now find length scale in clouds
+  cldthresh=0._r8
+  
+  ! determine the convective velocity scale at
+  !   the top of the cloud
+  
+  conv_vel(:,1)=0._r8
+  do k=2,nlev
+    do i=1,shcol
+      conv_vel(i,k) = conv_vel(i,k-1)+2.5_r8*dz_zt(i,k)*(ggr/basetemp)*wthv_sec(i,k)
+    enddo ! end i loop (column loop)
+  enddo ! end k loop (vertical loop)
+ 
+  if (doclouddef) then
+ 
+  do i=1,shcol
+  
+    if (cldarr(i) .eq. 1) then
+      
+      kl=0
+      ku=0
+      
+      do k=2,nlev-2
+        
+	! Look for cloud top in this column
+	if (cldin(i,k) .gt. cldthresh .and. kl .eq. 
```

Comment:
@bogensch , what assumptions are made regarding the contents of the inputs here?